protocol/packages/currencies/build/currencies_tree.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    borrow::Borrow,
+    collections::{
+        btree_map::{self, BTreeMap},
+        btree_set::{self, BTreeSet},
+    },
+    ops::ControlFlow,
+};
+
+use anyhow::{anyhow, Context as _, Result};
+
+use topology::{swap_pairs::PairTargets, Topology};
+
+use crate::protocol::Protocol;
+
+pub(crate) struct CurrenciesTree<'parents_of, 'parent, 'children_of, 'child> {
+    parents: BTreeMap<&'parents_of str, BTreeSet<&'parent str>>,
+    children: BTreeMap<&'children_of str, BTreeSet<&'child str>>,
+}
+
+impl<'topology> CurrenciesTree<'topology, 'topology, 'topology, 'topology> {
+    pub fn new(
+        topology: &'topology Topology,
+        protocol: &Protocol,
+        host_currency_ticker: &str,
+    ) -> Result<Self> {
+        let result = topology
+            .network_dexes(&protocol.dex_network)
+            .context("Selected DEX network doesn't define any DEXes!")?
+            .get(&protocol.dex)
+            .context("Selected DEX network doesn't define such DEX!")?
+            .swap_pairs()
+            .iter()
+            .map(|(ticker, targets)| (ticker.borrow(), targets))
+            .filter(|&(ticker, _)| {
+                super::filter_selected_currencies(protocol, host_currency_ticker, ticker)
+            })
+            .try_fold(
+                const {
+                    Self {
+                        parents: const { BTreeMap::<_, BTreeSet<_>>::new() },
+                        children: const { BTreeMap::<_, BTreeSet<_>>::new() },
+                    }
+                },
+                |currencies_tree, (ticker, targets)| {
+                    currencies_tree.process_targets(protocol, host_currency_ticker, ticker, targets)
+                },
+            );
+
+        match result {
+            ControlFlow::Continue(currencies_tree) => Ok(currencies_tree),
+            ControlFlow::Break(()) => Err(anyhow!(
+                "Currency ticker duplication detected in swap pairs!"
+            )),
+        }
+    }
+
+    fn process_targets(
+        mut self,
+        protocol: &Protocol,
+        host_currency_ticker: &str,
+        ticker: &'topology str,
+        targets: &'topology PairTargets,
+    ) -> ControlFlow<(), Self> {
```

Comment:
Let's avoid misusing that type in scenarios it is not designed for. After all, here we do not control any flow.